packages/core/data-transfer/src/directory/providers/source/index.ts

```diff
@@ -0,0 +1,208 @@
+
+/**
+ * Provider options
+ */
+export interface ILocalDirectorySourceProviderOptions {
+  directory: {
+    path: string; // the directory to load files from
+  };
+}
+
+export const createLocalDirectorySourceProvider = (
+  options: ILocalDirectorySourceProviderOptions
+) => {
+  return new LocalDirectorySourceProvider(options);
+};
+
+class LocalDirectorySourceProvider implements ISourceProvider {
+  type: ProviderType = 'source';
+
+  name = 'source::local-directory';
+
+  options: ILocalDirectorySourceProviderOptions;
+
+  #metadata?: IMetadata;
+
+  constructor(options: ILocalDirectorySourceProviderOptions) {
+    this.options = options;
+  }
+
+  /**
+   * Pre flight checks to ensure the directory exists and can be read.
+   */
+  async bootstrap() {
+    const { path: directoryPath } = this.options.directory;
+
+    try {
+      // Check if the directory exists
+      const stat = await fs.stat(directoryPath);
+      if (!stat.isDirectory()) {
+        throw new Error(`Path is not a directory: ${directoryPath}`);
+      }
+      // Load the metadata file
+      await this.#loadMetadata();
+    } catch (e) {
+      throw new ProviderInitializationError(
+        `Could not initialize provider: ${(e as Error).message}`
+      );
+    }
+
+    if (!this.#metadata) {
+      throw new ProviderInitializationError('Could not load metadata from the directory.');
+    }
+  }
+
+  async #loadMetadata() {
+    const metadataPath = path.join(this.options.directory.path, METADATA_FILE_PATH);
+    this.#metadata = await this.#readJSONFile<IMetadata>(metadataPath);
+  }
+
+  async #loadAssetMetadata(filePath: string) {
+    return this.#readJSONFile<IFile>(filePath);
+  }
+
+  async getMetadata() {
+    if (!this.#metadata) {
+      await this.#loadMetadata();
+    }
+
+    return this.#metadata ?? null;
+  }
+
+  async getSchemas() {
+    const schemaCollection = await utils.stream.collect<Struct.Schema>(
+      this.createSchemasReadStream()
+    );
+
+    if (isEmpty(schemaCollection)) {
+      throw new ProviderInitializationError('Could not load schemas from the directory.');
+    }
+
+    // Group schema by UID
+    const schemas = keyBy('uid', schemaCollection);
+
+    // Transform to valid JSON
+    return utils.schema.schemasToValidJSON(schemas);
+  }
+
+  createEntitiesReadStream(): Readable {
+    return this.#streamJsonlDirectory('entities');
+  }
+
+  createSchemasReadStream(): Readable {
+    return this.#streamJsonlDirectory('schemas');
+  }
+
+  createLinksReadStream(): Readable {
+    return this.#streamJsonlDirectory('links');
+  }
+
+  createConfigurationReadStream(): Readable {
+    return this.#streamJsonlDirectory('configuration');
+  }
+
+  createAssetsReadStream(): Readable {
+    const outStream = new PassThrough({ objectMode: true });
+    const assetsDir = path.join(this.options.directory.path, 'assets/uploads');
+
+    fs.readdir(assetsDir, { withFileTypes: true }, (err, files) => {
```

Comment:
Could be simpler to create a stream from an async generator and iterate through its content (e.g. using `fs.opendir`) rather than relying on a custom impl, what do you think ?